packages/cli/generators/repository/index.js

```diff
@@ -0,0 +1,398 @@
+    let modelFile = path.join(
+      this.artifactInfo.modelDir,
+      modelName.toLowerCase() + '.model.ts',
+    );
+    try {
+      fileContent = this.fs.read(modelFile, {});
+    } catch (err) {
+      debug(`${ERROR_READING_FILE} ${modelFile}: ${err.message}`);
+      return this.exit(err);
+    }
+
+    return tsquery.getIdFromModel(fileContent);
+  }
+
+  /**
+   * helper method to inspect and validate a repository type
+   */
+  async _inferRepositoryType() {
+    if (!this.artifactInfo.dataSourceClassName) {
+      return;
+    }
+    let result = this._isConnectorOfType(
+      KEY_VALUE_CONNECTOR,
+      this.artifactInfo.dataSourceClassName,
+    );
+    debug(`KeyValue Connector: ${result}`);
+
+    if (result) {
+      this.artifactInfo.repositoryTypeClass = KEY_VALUE_REPOSITORY;
+      this.artifactInfo.defaultTemplate = REPOSITORY_KV_TEMPLATE;
+    } else {
+      this.artifactInfo.repositoryTypeClass = DEFAULT_CRUD_REPOSITORY;
+      this.artifactInfo.defaultTemplate = REPOSITORY_CRUD_TEMPLATE;
+    }
+
+    // assign the data source name to the information artifact
+    let dataSourceName = this.artifactInfo.dataSourceClassName
+      .replace('Datasource', '')
+      .toLowerCase();
+
+    Object.assign(this.artifactInfo, {
+      dataSourceName: dataSourceName,
+    });
+  }
+
+  /**
+   * load the connectors available and check if the basedModel matches any
+   * connectorType supplied for the given connector name
+   * @param {string} connectorType single or a comma separated string array
+   */
+  _isConnectorOfType(connectorType, dataSourceClassName) {
+    debug(`calling isConnectorType ${connectorType}`);
+    let jsonFileContent = '';
+    let result = false;
+
+    if (!dataSourceClassName) {
+      return false;
+    }
+    let datasourceJSONFile = path.join(
+      this.artifactInfo.datasourcesDir,
+      dataSourceClassName
+        .replace('Datasource', '.datasource.json')
+        .toLowerCase(),
+    );
+
+    try {
+      jsonFileContent = this.fs.readJSON(datasourceJSONFile, {});
+    } catch (err) {
+      debug(`${ERROR_READING_FILE} ${datasourceJSONFile}: ${err.message}`);
+      return this.exit(err);
+    }
+
+    for (let connector of Object.values(connectors)) {
+      const matchedConnector =
+        jsonFileContent.connector === connector.name ||
+        jsonFileContent.connector === `loopback-connector-${connector.name}`;
+
+      if (matchedConnector && connectorType.includes(connector.baseModel)) {
+        result = true;
+        break;
+      }
+    }
+
+    return result;
+  }
+
+  _setupGenerator() {
+    this.artifactInfo = {
+      type: 'repository',
+      rootDir: 'src',
+    };
+
+    this.artifactInfo = {
+      type: 'datasource',
+      rootDir: 'src',
+    };
+
+    this.artifactInfo.outDir = path.resolve(
+      this.artifactInfo.rootDir,
+      'repositories',
+    );
+    this.artifactInfo.datasourcesDir = path.resolve(
+      this.artifactInfo.rootDir,
+      'datasources',
```

Comment:
It would be great if we could share the code building `datasourcesDir`, `modelDir` and other convention-based directories between generators, so that there is a single place defining our convention.

Reply:
for now, these are only used to read the ts files in the specified directory.  you mean having these definitions as export members in a library ?

Reply:
I mean that `lb4 datasource` must be using this same approach based on the magic string `datasources` to find the directory where to put the newly generated `.ts` and `.json` files. Maybe it's handled by one of the base generator classes?

Ideally, the code determining where to write a file like `src/datasource/db.datasource.ts`  should be the same that's used to determine where to look for those datasource files.

I am not sure if I have explained myself well, please do ask more questions if needed!

At the same time, this is not a big deal in the context of this pull request, we can defer such cleanup for later, after this patch is landed.

Reply:
> ok, one single place to define the directory locations and then import them from all generators , correct?

Yes, exactly!

Reply:
I think we should address this in a separate PR and consolidate in one place these feature for all generators, what do you think?

Reply:
Sure, I am fine with addressing this in a separate PR.

See my earlier comment above.
> At the same time, this is not a big deal in the context of this pull request, we can defer such cleanup for later, after this patch is landed.